library/L1_Peripheral/lpc40xx/gpio.hpp

```diff
@@ -22,258 +26,319 @@ class Gpio final : public sjsu::Gpio
+      }
+      else if constexpr (IsPlatform(sjsu::build::Platform::lpc40xx))
+      {
+        pin_obj_ = &kLpc17xxPin;
+      }
+      else
+      {
+        pin_obj_ = &sjsu::GetInactive<sjsu::Pin>();
+      }
+    }
+    else
+    {
+      pin_obj_ = pin;
+    }
+
+    // Assign interrupt_index_ pointer
+    if (port_number == 2)
+    {
+      interrupt_index_ = 1;
+    }
+    else if (port_number == 0)
+    {
+      interrupt_index_ = 0;
+    }
+    else
+    {
+      interrupt_index_ = kInterruptPorts;
+    }
+
+    // Assign gpio port
+    gpio_port_ = *GpioRegister(port_number);
   }
   void SetDirection(Direction direction) const override
   {
-    pin_.SetPinFunction(kGpioFunction);
-    volatile uint32_t * dir_register = &gpio_port[pin_.GetPort()]->DIR;
+    /// Pin function is zero fall pins on the LPC40xx and LPC17xx.
+    constexpr uint8_t kGpioFunction = 0;
+    pin_obj_->SetPinFunction(kGpioFunction);
 
     if (direction == Direction::kInput)
     {
-      *dir_register = bit::Clear(*dir_register, pin_.GetPin());
+      gpio_port_->DIR = bit::Clear(gpio_port_->DIR, pin_);
     }
     else
     {
-      *dir_register = bit::Set(*dir_register, pin_.GetPin());
+      gpio_port_->DIR = bit::Set(gpio_port_->DIR, pin_);
     }
   }
   void Set(State output = kHigh) const override
   {
     if (output == State::kHigh)
     {
-      gpio_port[pin_.GetPort()]->SET = (1 << pin_.GetPin());
+      gpio_port_->SET = (1 << pin_);
     }
     else
     {
-      gpio_port[pin_.GetPort()]->CLR = (1 << pin_.GetPin());
+      gpio_port_->CLR = (1 << pin_);
     }
   }
   void Toggle() const override
   {
-    gpio_port[pin_.GetPort()]->PIN ^= (1 << pin_.GetPin());
+    gpio_port_->PIN ^= (1 << pin_);
   }
   bool Read() const override
   {
-    return bit::Read(gpio_port[pin_.GetPort()]->PIN, pin_.GetPin());
+    return bit::Read(gpio_port_->PIN, pin_);
   }
   const sjsu::Pin & GetPin() const override
   {
-    return pin_;
+    return *pin_obj_;
   }
 
-  /// Checks if the selected gpio port is valid for external interrupts.
-  bool ValidPortCheck() const
+  /// Assign the developer's ISR and sets the selected edge that the gpio
+  /// interrupt will be triggered on.
+  void AttachInterrupt(InterruptCallback callback, Edge edge) override
   {
-    bool is_valid = (interupt_port_ <= 1);
-    SJ2_ASSERT_WARNING(is_valid,
-                       "Port %d cannot be used for External Interrupts. Need "
-                       "to use GPIO on Port 0 or 2.",
-                       pin_.GetPort());
-    return is_valid;
-  }
+    if (!IsAValidPort())
+    {
+      return;
+    }
 
-  /// Assigns the developer's ISR function to the port/pin gpio instance.
-  void SetInterruptRoutine(InterruptCallback callback) const
-  {
-    ValidPortCheck();
-    interrupt_handlers[interupt_port_][pin_.GetPin()] = callback;
-  }
+    sjsu::InterruptController::GetPlatformController().Enable({
+        .interrupt_request_number = GPIO_IRQn,
```

Comment:
A platform check may be needed to get the interrupt number. For LPC17xx there is no `GPIO_IRQn`. From the data sheet in 9.2.2, GPIO0 and GPIO2 interrupts share the same NVIC with `EINT3_IRQn`.

Reply:
Ooooo! Good catch. I'll make the change.